Style: Reduce size of code font and improve responsiveness at narrow screen widths

CAM-Gerlach opened the pull request:
Right now, the code font is much larger than the prose font, which varies by platform (due to the former using a system font stack, while the latter uses a web font). As a stopgap solution, to ensure 79 characters fit on a line under all but the most constrained screen resolutions, and reduce the perceptible size difference between the prose and code fonts while making it more motivated, this PR follows @pradyunsg 's suggestion reduce the code font size to something similar to what Furo uses, 0.875rem/14px as was suggested and discussed there, until we decide on a final resolution to the font stack issue.

Furthermore, as also suggested by @methane on #2347 , it allows code blocks to break by letter than than by word, resolving formatting issues at narrow screen widths and matching the behavior of most editors/IDEs (while ensuring inline code breaks cleanly).

Resolves #2437
Resolves #2449

methane commented on pep_sphinx_extensions/pep_theme/static/style.css:
```diff
@@ -121,6 +122,7 @@ pre {
 /* Contents rules */
 details > summary {
     font-weight: bold;
+    word-break: break-all;
```
Is `details > summary` a right place?

CAM-Gerlach replied:
Wow, that's embarrassing...I meant to put it in the `pre` block right above, but wasn't paying as much attention as I should have (since I was testing it live in the browser dev tools rather than waiting for a full rebuild). My mistake; fixed now.

methane reviewed (approved):
Looks OK on macOS and Pixel 3.

CAM-Gerlach commented:
I've made some further improvements here based on additional feedback on #2437 . See [my comment there](https://github.com/python/peps/issues/2437#issuecomment-1070321772) for more details.

pradyunsg commented:
This isn't specific to this PR, but we might want to reconsider making the table font smaller as the screen width reduces.

<p align="center">
<img src="https://user-images.githubusercontent.com/3275593/158759228-f842630d-235e-4474-85c1-da5dca8181ac.png" alt="screenshot" width="380">
</p>

hugovk commented:
I agree, changes here generally look good but let's not make the table text smaller, it makes is harder to read. If anything, we should be increasing the font size and whitespace on mobile.

Plus smaller text doesn't help for wider tables.

For example, the wide table in PEP 594 still affects the whole zoom on the page (Android/Samsung S10/Chrome):

<img width=50% src=https://user-images.githubusercontent.com/1324225/158775516-0847a922-98dd-4a93-8f7d-a679381f8fa4.png>

* https://pep-previews--2439.org.readthedocs.build/pep-0594/#deprecated-modules
* https://peps.python.org/pep-0594/#deprecated-modules

There is [a whole bunch of things](https://adrianroselli.com/2017/11/a-responsive-accessible-table.html) that can be done to make tables responsive and accessible, but:

> ... the simplest way to make a table responsive is to put it in a container that will scroll. This way it does not break your design by pushing the layout and content past the viewport edges.

As a quick example, [Boostrap does this](https://getbootstrap.com/docs/4.0/content/tables/#responsive-tables) too:

> Responsive tables allow tables to be scrolled horizontally with ease.

<img width=50% src=https://user-images.githubusercontent.com/1324225/158777540-3ce5ab87-8e98-48cd-9431-88e0655a81bb.png>

Can we do the same?

pradyunsg commented:
That requires rewriting HTML that is generated by Sphinx (or doing a transform within Sphinx). There's code for that in various Sphinx themes that can be borrowed for this (for Furo's code, see https://github.com/pradyunsg/furo/blob/812935eee54aeec3ae964647c37f2f18d930c9ec/src/furo/__init__.py#L42).

pradyunsg commented:
Or, that's the only way that I know of. If you know of a better way, please say so here. I'd be more than happy to eliminate an HTML parse of potentially long documents from Furo's rendering step. :)

CAM-Gerlach commented:
To address @sfermigier 's feedback on #2437 as well as my own observations (and in line with the comments here about increasing whitespace in general), I've increased the padding a bit around the main content container at narrower width breakpoints to avoid it being too cramped and not having any margin to the screen edge.

> For example, the wide table in PEP 594 still affects the whole zoom on the page (Android/Samsung S10/Chrome):

Hmm; I didn't fully optimize for widths lower than about 380-390 px; it looks like the S10 series has a viewport width of 360 px. Honestly, I don't understand why manufacturers use such a crazy-high DPR on such devices; it seems like such a waste of its huge 6.1-6.7 inch 1440p screen.

In any case, while I do see a little cropping on my simulated S10 (via FF dev tools), I don't see nearly the degree you do, so not sure how to repro...ultimately, without HTML-based solutions or direct control over the content (we can't exactly tell authors they can only use skinny tables width to cater to certain very narrow-viewport mobile devices), its going to be a "good-enough" approach as a stopgap that won't nessesarily work perfectly for every table on every PEP on every device.

<details>

<summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/17051931/158864211-ca3493ae-804d-41d5-b739-79fca37e5132.png)

</details>

> That requires rewriting HTML that is generated by Sphinx (or doing a transform within Sphinx) to wrap the table tag in a div. There's code for that in various Sphinx themes that can be borrowed for this (for Furo's code which reparses the generated HTML, see https://github.com/pradyunsg/furo/blob/812935eee54aeec3ae964647c37f2f18d930c9ec/src/furo/__init__.py#L42).

Thanks. I agree something like that is probably the way to go (ideally using our existing HTML translation extensions rather than via post-build hacks). However, I'm not sure if that's really a good idea to implement in this PR focused just on straightforward CSS style tweaks to fix the immediate issues, since it is kinda hacky, entails significant code changes to the rendering system, and requires adding at least one new top-level dependency (BeautifulSoup).

Also, it would greatly benefit from the input of our lead rendering system maintainer @AA-Turner , who is taking a well-deserved several-week break for academic reasons after getting the new rendering system across the immediate finish line, as well as additional time to plan, prototype and discuss the best approach.

As such, I'd recommend we defer that to a followup, and falling back here to the least-bad approach that can be realized in CSS.  I realize shrinking the table font is not ideal as it decreases readability while not scaling all the way to the narrowest widths, but ultimately potentially requiring some users who want to read the particulars of a table but have trouble doing so to zoom in on tables is ultimately a lot less disruptive than requiring all users zoom out or scroll to see all of the rest of the page content, right?

Its certainly possible that there's a better stopgap solution that can be implemented quickly without requiring major changes, but one major constraint is that we don't really control the content, and more drastic changes like stacking can harm usability and subvert author expectations for how their content is displayed. Thoughts?

hugovk commented:
> > That requires rewriting HTML that is generated by Sphinx (or doing a transform within Sphinx) to wrap the table tag in a div. There's code for that in various Sphinx themes that can be borrowed for this (for Furo's code which reparses the generated HTML, see [pradyunsg/furo@`812935e`/src/furo/__init__.py#L42](https://github.com/pradyunsg/furo/blob/812935eee54aeec3ae964647c37f2f18d930c9ec/src/furo/__init__.py#L42)).
> 
> Thanks. I agree something like that is probably the way to go (ideally using our existing HTML translation extensions rather than via post-build hacks). However, I'm not sure if that's really a good idea to implement in this PR focused just on straightforward CSS style tweaks to fix the immediate issues, since it is kinda hacky, entails significant code changes to the rendering system, and requires adding at least one new top-level dependency (BeautifulSoup).

Short term: if you want this PR to be CSS only, shall we remove the table shrinking CSS from this PR and do the Furo trick in another PR?

Medium term: is it worth asking Sphinx to wrap tables in a div? That would benefit us, Furo users and other Sphinx users.

Longer term: when Adam's back we can discuss further improvements in more depth.

pradyunsg commented:
> Medium term: is it worth asking Sphinx to wrap tables in a div? That would benefit us, Furo users and other Sphinx users.

I think it is. I'm not quite sure how Sphinx handles markup changes under their somewhat-strict backwards compatibility model -- but yes, we should totally ask them! :)

CAM-Gerlach commented:
> Short term: if you want this PR to be CSS only, shall we remove the table shrinking CSS from this PR and do the Furo trick in another PR?

Well, if we do that, it won't fix the primary issue reported in #2437 , which motivated this PR in the first place, and which the users who reported the issue and tested this seemed to be happy with. As I mentioned above,

> I realize shrinking the table font is not ideal as it decreases readability while not scaling all the way to the narrowest widths, but ultimately potentially requiring some users to zoom in on a specific table is ultimately a lot less disruptive than requiring all users zoom out or scroll to see all of the rest of the page content, right?

It seems to be at least a modest net improvement for the majority of cases for the time being until a followup PR with a longer-term solution is implemented, particularly if we defer the Furo hack, since it will require an extra round of discussion, implementation in and testing to hack a kludge in, as well as additional dependencies, that will almost certainly get ripped back out again when the earlier of A. Adam gets back in a month and we implement it in the HTML translator, or B. Sphinx wraps the table in a div.

CAM-Gerlach commented:
What about as a lighter-weight stopgap, we just wrap body tables in a `<div>` using a JS snippit? This should achieve the same result as the Furo hack while avoiding most of the costs, as it doesn't require changing the build system, pulling in extra deps and hacking the raw HTML. I'm pretty sure that's how I've handled similar situations to this in the past, IIRC, and It should be simple enough that I could just do it in this PR.

hugovk commented:
Okay, let's try some JavaScript. In general, I think it's better to fix the HTML itself, for those without JavaScript, but hopefully this a temporary thing!

CAM-Gerlach commented:
Yeah, better to fix the HTML on the Sphinx side, but indeed this JS will only be a temporary solution.

In any case, I've implemented it as described above (via JS using `overflow-x`) and ditched the previous font scaling approach.

> for those without JavaScript,

Well, FWIW, considering [98% of websites use JS](https://w3techs.com/technologies/details/cp-javascript), only around [0.2% of users in 2016 had it disabled at some level](https://deliberatedigital.com/blockmetry/javascript-disabled), and this is only a non-essential usability enhancement, I'm not sure how compelling that reason is, particularly considering [over 10 times that (1.5-3.6%)](https://caniuse.com/viewport-units) have browsers ancient enough that don't support the `vw` units currently used in the pure-CSS solution, and similar for other features of the existing CSS and JS.

hugovk commented on pep_sphinx_extensions/pep_theme/static/style.css:
```diff
@@ -177,6 +177,11 @@ sup {top: -0.5em}
 /* Table rules */
+div.table-wrapper {
+    overflow-x: scroll;
+    scrollbar-width: thin;  /* CSS Standards, not *yet* widely supported */
```
Indeed, only supported by Firefox: https://caniuse.com/?search=scrollbar-width

MDN also notes accessibility concerns, `scrollbar-width: thin` "can make content hard or impossible to scroll if the author does not provide an alternative scrolling mechanism": https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-width#accessibility_concerns

As we're doing this partly for accessibility shall we skip this one?

CAM-Gerlach replied:
Yeah, I wanted to be consistent with the other scroll bars (e.g. for the sidebar), but that also uses the illegal, bespoke pseudo-elements to support WebKit/Blink-based browsers, so you're right—best to just eliminate this, fall back on the device-default scrollbars and use `overflow-x: auto` instead to automatically hide it when not needed instead of a transparent overlay. I also fixed several other additional existing responsiveness issues I spotted in further testing for this.

hugovk reviewed (commented):
Looking good, thank you!

One comment here.

And I found something else needing a break, but I'll raise that separately so as not to delay this any more.

CAM-Gerlach commented:
I eliminated the poorly-supported, potentially inaccessible scroll bar tweaks as @hugovk suggested, and switched it to `auto` mode instead to hide and show the scroll bar on demand. I also fixed a couple more outstanding layout/responsiveness issues I spotted in further testing this, including:

* Allowing breaking by letter in footnotes to avoid them also causing scrolling and sub-par layouts at very narrow widths, due to long URLs
* Fixing a significant issue (currently present in production) where the sidebar was displayed _below_ all of the content rather than hidden entirely at moderately narrow widths (when we already have it under the `Contents` dropdown at the top)
* Adjusting the sidebar breakpoint to still show it for larger tablets, and modern phones in landscape mode, where there's room for it and it can ease navigation, while moving some of the other responsive tweaks up a little further to fall cleanly between these device classes
* Eliminating the other instance of illegal scrollbar hacks that are browser dependent, poorly supported and harm accessibility, to be consistent with the requested changes to the table scrollbar above,

hugovk commented:
Looking good!

> * Allowing breaking by letter in footnotes to avoid them also causing scrolling and sub-par layouts at very narrow widths, due to long URLs

Nice, this was the other thing I was going to bring up :)

Breaking by word works nicely for URLs too (`break-all` left, `break-word` right):

![image](https://user-images.githubusercontent.com/1324225/159425778-f0c04142-b053-4bba-a78a-403236975682.png)

https://pep-previews--2439.org.readthedocs.build/pep-0594/

And some PEPs have a lot of footnote text where `break-all` is inappropriate (again `break-all` left, `break-word` right):

![image](https://user-images.githubusercontent.com/1324225/159426184-e289a3a7-d1a2-42e7-8c4b-e0db7d7c009b.png)

https://pep-previews--2439.org.readthedocs.build/pep-0465/

The same PEP also has some long URLs in the body, causing a pesky horizontal scroll. Left is PR, right is with `word-break: break-word` on `<body>`:

![image](https://user-images.githubusercontent.com/1324225/159426988-e2745bdc-5b20-4966-aae3-19465605f7f4.png)

CAM-Gerlach commented:
Hmm, I see. To note, I couldn't reproduce the issue in your screenshots (at least on FF on Windows), except at much narrower widths, because by default links _were_ breaking on `/`s for me. Furthermore, applying said solution also caused the footnote numbers to render very poorly, one character to a line. What's more, `work-break: break-word` is deprecated, and only supported in certain browsers, and so is the `word-wrap` property, which was previously used for this. 

To resolve all this, I used `overflow-wrap: anywhere` (with a fallback to `overflow-wrap: break-word` for the usual suspect, Webkit, that doesn't support modern standards), which achieves similar results to `word-break: break-word` in a modern, standards-conformant way. This ensures that breaks are only used when they are truly needed to prevent an overflow, and inline lines aren't broken arbitrarily, while still covering all links. Also, I fixed a padding issue that avoided the breaks in the footnote numbers. This should solve the problem once and for all, at least at all modern device widths. 

(To note, neither of these cases (long links in the body, nor footnotes in general) would be a real problem for new or updated PEPs that followed the current guidance proscribing bare and footnote links in favor of inline links with proper display text, and this rule would be unnecessary.)

hugovk commented:
Thanks for all this, I think we're good now!

---

I should have mentioned my screenshots were Chrome/macOS.

Yep, good idea to use `overwrap: anywhere` instead of deprecated `workd-break: break-word`. [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/word-break) suggests that as a replacement.

Here's https://pep-previews--2439.org.readthedocs.build/pep-0465/ on Chrome with mobile layout.

# Chrome

![image](https://user-images.githubusercontent.com/1324225/159475252-e5e20d6d-8728-4c9c-b8aa-6f9971815593.png)

# Firefox

![image](https://user-images.githubusercontent.com/1324225/159475463-58f429f6-6048-4bb0-8a40-667839468b12.png)

# Safari

![image](https://user-images.githubusercontent.com/1324225/159476116-bff4ac9c-430e-4370-9139-6591eb9533c8.png)

---

> To resolve all this, I used `overflow-wrap: anywhere` (with a fallback to `overflow-wrap: break-word` for the usual suspect, Webkit, that doesn't support modern standards), which achieves similar results to `word-break: break-word` in a modern, standards-conformant way. This ensures that breaks are only used when they are truly needed to prevent an overflow, and inline lines aren't broken arbitrarily, while still covering all links.

You'll see from the inspector that all three show:

~overflow-wrap: break-word;~
overflow-wrap: anywhere;

https://caniuse.com/wordwrap shows widespread support. Anyway, fine to keep both if you wish :)

hugovk commented:
I'll merge this now, thanks again!

CAM-Gerlach commented:
Hmm, [CanIUse](https://caniuse.com/mdn-css_properties_overflow-wrap_anywhere), using data from on [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-wrap#browser_compatibility), claims that Safari doesn't have support for `overflow-wrap: anywhere`, but it seems that may be out of date, at least in recent versions. Either way, since it isn't non-standard, I don't think it hurts to keep it as a fallback for now.

encukou commented:
Wrapping makes the `# These lines are`... comments in [PEP 678](https://peps.python.org/pep-0678/#raise-wrapper-explanation-from-err) unreadable on smaller screens.

hugovk commented:
Desktop:

![image](https://user-images.githubusercontent.com/1324225/159686614-37df8ab5-8d2a-41ed-8245-4232738a9926.png)

Mobile:

![image](https://user-images.githubusercontent.com/1324225/159686707-5f44ea99-d172-403e-b0f6-1cae25061946.png)

(Chrome/macOS)

Instead of wrapping, perhaps we should do the horizontal scroll thing on code, like with tables?

GitHub does horizontal scroll for code (files, diffs, and in READMEs etc).

CAM-Gerlach commented:
Thanks; that makes more sense (at least for the great majority of literal blocks), and I'm not sure why I didn't think of it before. To note, readability is further impaired by that block not having the correct syntax highlighting language tag applied, so syntax highlighting doesn't work; I'll go through and fix that separately in a PR for that purpose.

I've opened #2463 which implements the suggested solution, using `overflow-x: auto` instead of `overflow-wrap: anywhere` on `pre` blocks, and switching them from `white-space: pre-wrap` to `white-space: pre`.

hugovk commented:
https://github.com/python/peps/pull/2463 is merged! 🚀
